prime-router/src/main/kotlin/fhirengine/engine/FHIRConverter.kt

```diff
@@ -175,6 +199,252 @@ class FHIRConverter(
         return emptyList()
     }
 
+    data class ProcessedItem<ParsedType>(
+        val rawItem: String,
+        val index: Int,
+        val parsedItem: ParsedType? = null,
+        val parseError: ActionLogDetail? = null,
+        val validationError: ActionLogDetail? = null,
+        val bundle: Bundle? = null,
+    ) {
+        fun updateParsed(error: ActionLogDetail?): ProcessedItem<ParsedType> {
+            return this.copy(parseError = error)
+        }
+
+        fun updateParsed(parsed: ParsedType): ProcessedItem<ParsedType> {
+            return this.copy(parsedItem = parsed)
+        }
+
+        fun updateValidation(error: ActionLogDetail): ProcessedItem<ParsedType> {
+            if (parseError == null && parsedItem != null) {
+                return this.copy(validationError = error)
+            }
+            throw RuntimeException("Validation should not be set since item was not parseable")
+        }
+
+        fun setBundle(bundle: Bundle): ProcessedItem<ParsedType> {
+            if (parseError == null && validationError == null) {
+                return this.copy(bundle = bundle)
+            }
+            throw RuntimeException("Bundle should not be set if the item was not parseable or valid")
+        }
+
+        fun getError(): ActionLogDetail? {
+            return parseError ?: validationError
+        }
+    }
+
+    private fun measureProcessingItems(
+        reportId: UUID,
+        format: Report.Format,
+        getProcessedItems: () -> List<ProcessedItem<*>>,
+    ): List<ProcessedItem<*>> {
+        val timeSource = kotlin.time.TimeSource.Monotonic
+        timeSource.markNow()
+        val items = getProcessedItems()
+        val end = timeSource.markNow()
```

Comment:
I believe you're using the time api wrong here. You don't need to save the end timestamp. A call to `elapsedNow` does the operation under the hood.

```kotlin
val now = timeSource.markNow()
val items = getProcessedItems()
val elapsed = now.elapsedNow()
```

Even more simply for something small like this:
```kotlin
val items: List<ProcessedItem<*>>
val elapsed = measureTime {
    items = getProcessedItems()
}
```